qualtran/bloqs/chemistry/trotter/hubbard/qpe_cost_optimization.ipynb

```diff
@@ -0,0 +1,574 @@
+    "\n",
+    "delta_ht_opt, delta_ts_opt, delta_pe_opt, t_opt = minimize_linesearch(n_rot, n_t, xi_bound, prod_ord)\n",
+    "print(rf\"\\Delta_{{HT}} = {delta_ht_opt}\")\n",
+    "print(rf\"\\Delta_{{TS}} = {delta_ts_opt}\")\n",
+    "print(rf\"\\Delta_{{PE}} = {delta_pe_opt}\")\n",
+    "print(rf\"T_{{opt}} = {t_opt:4.3e}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Symbolic T counts\n",
+    "\n",
+    "We can avoid doing any manipulation ourselves using sympy to represent the error expression."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "s_eps_r, s_length, s_hubb_u, s_timestep, s_tau = sympy.symbols(r'\\epsilon_{R}, L, u, t, \\tau')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "First let's check the Bloq counts look correct for the Trotter step, there are two sources rotations from the interaction and hopping bloq and some direct T gates from the `TwoBitFFFT` gate."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from qualtran.resource_counting.t_counts_from_sigma import t_counts_from_sigma\n",
+    "\n",
+    "s_trotter_step = build_plaq_unitary_second_order_suzuki(\n",
+    "    s_length, s_hubb_u, s_timestep, eps=s_eps_r, hubb_t=s_tau\n",
+    ")\n",
+    "t_counts = t_counts_from_sigma(s_trotter_step.call_graph(generalizer=catch_rotations)[1])\n",
+    "t_counts"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# check the symbolic counts match the expected counts\n",
+    "t_counts_orig = t_counts_from_sigma(trotter_step.call_graph(generalizer=catch_rotations)[1])\n",
+    "# for some reason substituting both at once leads to a precision error\n",
+    "t_counts = t_counts.evalf(subs={s_eps_r: eps_single_rot})\n",
+    "t_counts_symb = t_counts.evalf(subs={s_length: length})\n",
+    "assert t_counts_orig == t_counts_symb"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now let's reproduce the expression for the total cost"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "s_delta_ht, s_delta_ts, s_delta_pe, s_p, s_xi = sympy.symbols(\n",
+    "    '\\Delta_{HT}, \\Delta_{TS}, \\Delta_{PE}, p, xi'\n",
+    ")\n",
+    "s_n_rot, s_n_t, s_n_pe = sympy.symbols('N_R, N_T, N_PE')\n",
+    "s_timestep = (s_delta_ts / s_xi) ** (1 / s_p)\n",
+    "s_eps_r = (s_delta_ht * s_timestep) / s_n_rot\n",
+    "s_n_pe = 0.76 * sympy.pi / (s_delta_pe * s_timestep)\n",
+    "s_trotter_step = build_plaq_unitary_second_order_suzuki(\n",
+    "    s_length, s_hubb_u, s_timestep, eps=s_eps_r, hubb_t=s_tau\n",
+    ")\n",
+    "# just use this cost in lieu of a QPE bloq\n",
```

Comment:
In the paper they use adoptive phase estimation that uses a single control qubit but cost goes as `0.76 * pi`. We have the optimal scaling version where cost goes as `0.5 * pi` but it uses `m` control bits. We also have an open issue to implement https://arxiv.org/abs/2303.12503, which is the optimal scaling version but with only 2 control qubits. 

It'd be nice if we can either 
(a) Use a bloq (maybe just a shim for now) to get the cost for adoptive phase measurement based QPE used here so that we can add an implementation of the bloq later and keep track of the different QPE versions used and supported in cost estimates throughout the library (instead of a hardcoded formula)
(b) If (a) is too much work, then open an issue to do this at a later point and track that we are using a hardcoded formula instead of a bloq to get cost estimates for QPE here. 


I know doing (a) may require more than just a shim because of the new capabilities required in `SympySymbolAllocator` so I'm fine with (b) as well to make sure we can merge this PR quickly

Reply:
I think I'll add an issue for now, and add the shim as a follow up. I need to remind myself of the sympy allocation updates. It would also be nice to compare all three alternatives. I have some bandwidth for the next few weeks.

Reply:
Done.